### PR TITLE
Bug 1860922: add a controller to unmanage the OAuth API

### PR DESCRIPTION
**Pull request by stlaz**
Adds a controller to unmanage the OAuth API in case of a downgrade

/assign @p0lyn0mial

### CODE REVIEW

File: pkg/controller/apiservices/unmanageoauthapicontroller.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	operatorconfigclient "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+	operatorclientinformers "github.com/openshift/client-go/operator/informers/externalversions"
+	operatorlistersv1 "github.com/openshift/client-go/operator/listers/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// NewUnmanageAPIServicesController sets ManagingOAuthAPIServer flag to false.
+// This will make OAuth APIServer (introduced in 4.6) to step down and traffic will be routed to the OpenShift APIServer
```
Comment (p0lyn0mial) [inline pkg/controller/apiservices/unmanageoauthapicontroller.go:16]: Maybe comment that this is important for a downgrade 4.6 -> 4.5

File: pkg/controller/apiservices/unmanageoauthapicontroller_test.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+			}
+			fakeOperatorClient := fakeoperator.NewSimpleClientset(authOps...)
+			operatorLister := operatorv1listers.NewAuthenticationLister(indexer)
+
+			testRecorder := events.NewInMemoryRecorder("test")
+			if gotErr := syncUnmanageAPIServicesController(fakeOperatorClient.OperatorV1(), operatorLister)(context.TODO(), testSyncContext{recorder: testRecorder}); tt.expectErr != (gotErr != nil) {
+				t.Errorf("syncUnmanageAPIServicesController() => expected error: %v, but got %v", tt.expectErr, gotErr)
+			}
+
+			if tt.changed != (len(testRecorder.Events()) > 0) {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controller/apiservices/unmanageoauthapicontroller_test.go]: made it check both so that we don't post events if it's not necessary

**Comment by sdodson** [comment]
/hold
Since this becomes critical only when downgrading from 4.6 to 4.5 we're going to defer merging this until after we've promoted 4.4 to 4.5 upgrades to the stable channel. We'll consider this again in two weeks on August 12th.

**Comment by sttts** [comment]
>  We'll consider this again in two weeks on August 12th.

Aug 12th is fine. Note that QE cannot start testing https://issues.redhat.com/browse/MSTR-891 for 4.6 until this has merged. I.e. postponing it yet another z release would risk 4.6.

**Pull request merged**